lib/reek/spec/should_reek_of.rb

```diff
@@ -17,22 +23,42 @@ def initialize(smell_category,
 
       def matches?(actual)
         self.examiner = Examiner.new(actual, configuration: configuration)
-        self.all_smells = examiner.smells
-        all_smells.any? { |warning| warning.matches?(smell_category, smell_details) }
+
+        set_failure_messages_for_smell_type && return unless matching_smell_types?
+        set_failure_messages_for_smell_details && return unless matching_smell_details?
+
+        true
```

Comment:
> This code gives the impression that both validations are independent from each other. But they are not. The second one is dependent on the first one because it can't stand on it's own.

Oh ok, this is a very good point. My initial proposition (which I refactored to the above) was the below `case` – maybe this is a good compromise between readability, the right mechanics and clearer `#matches?` return values?

```ruby
def matches?(actual)
  self.examiner = Examiner.new(actual, configuration: configuration)
  case
  when not matching_smell_types?
    set_failure_messages_for_smell_type
    false
  when not matching_smell_details?
    set_failure_messages_for_smell_details
    false
  else
    true
  end
end
```

(This all stems from `#matches?` doing two distinct things – setting the failure messages and returning a proper boolean. The above at least makes it clear what’s going on when.)